server/src/main/java/org/opensearch/index/translog/transfer/FileTransferTracker.java

```diff
@@ -85,46 +99,118 @@ void add(String file, boolean success) {
-        });
+        updateTransferState(fileTransferTracker, file, targetState);
+    }
+
+    @Override
+    public void onSuccess(TranslogCheckpointSnapshot fileSnapshot) {
+        try {
+            long durationInMillis = (System.nanoTime() - fileTransferStartTime) / 1_000_000L;
+            remoteTranslogTransferTracker.addUploadTimeInMillis(durationInMillis);
+            updateTransferStats(fileSnapshot, true);
+        } catch (Exception ex) {
+            logger.error("Failure to update translog upload success stats", ex);
+        }
+        addGeneration(fileSnapshot.getGeneration(), TransferState.SUCCESS);
+        if (!ckpAsTranslogMetadata) {
+            add(fileSnapshot.getCheckpointFileName(), TransferState.SUCCESS);
+            add(fileSnapshot.getTranslogFileName(), TransferState.SUCCESS);
+        }
     }
 
     @Override
-    public void onFailure(TransferFileSnapshot fileSnapshot, Exception e) {
+    public void onFailure(TranslogCheckpointSnapshot fileSnapshot, Exception e) {
         long durationInMillis = (System.nanoTime() - fileTransferStartTime) / 1_000_000L;
         remoteTranslogTransferTracker.addUploadTimeInMillis(durationInMillis);
-        remoteTranslogTransferTracker.addUploadBytesFailed(bytesForTlogCkpFileToUpload.get(fileSnapshot.getName()));
-        add(fileSnapshot.getName(), TransferState.FAILED);
+        addGeneration(fileSnapshot.getGeneration(), TransferState.FAILED);
+
+        if (ckpAsTranslogMetadata) {
+            updateTransferStats(fileSnapshot, false);
+        } else {
+            assert e instanceof TranslogTransferException;
+            TranslogTransferException exception = (TranslogTransferException) e;
+            Set<TransferFileSnapshot> failedFiles = exception.getFailedFiles();
+            Set<TransferFileSnapshot> successFiles = exception.getSuccessFiles();
+            assert failedFiles.isEmpty() == false;
+            failedFiles.forEach(failedFile -> {
+                add(failedFile.getName(), false);
+                long failedBytes = 0;
+                try {
+                    failedBytes = failedFile.getContentLength();
+                } catch (IOException ignore) {}
+                updateBytesInRemoteTranslogTransferTracker(failedBytes, false);
+            });
+            successFiles.forEach(successFile -> {
+                add(successFile.getName(), true);
+                long succededBytes = 0;
+                try {
+                    succededBytes = successFile.getContentLength();
+                } catch (IOException ignore) {}
+                updateBytesInRemoteTranslogTransferTracker(succededBytes, true);
+            });
+        }
     }
 
-    public void delete(List<String> names) {
+    private void updateTransferStats(TranslogCheckpointSnapshot fileSnapshot, boolean isSuccess) {
+        Long translogFileBytes = bytesForTlogCkpFileToUpload.get(fileSnapshot.getTranslogFileName());
+        if (translogFileBytes != null) {
```

Comment:
when can this be null?

Reply:
In the w/o metadata case. When one of the file is already uploaded we will not store it bytes in the `bytesForTlogCkpFileToUpload` for that file, in that case `translogFileBytes` can be null.